### ISSUE TITLE
Error Stack as rewritten by Zone.js not recognized by Stackdriver

### ISSUE THREAD

**Issue by 4oo** [angular/angular#31588]
# 🐞 bug report

### Affected Package
The issue is caused by the zone.js package

### Is this a regression?
Not to my knowledge.

### Description

TL;DR Zone.js rewrites the error stack in a way that stackdriver doesn't recognize and event reports are bounced with 400 errors.

Zone.js rewrites the `stack` property on errors to be zone aware:
https://github.com/angular/angular/blob/3166cffd28e12d882bef3e55472e0a50f4cfc6f5/packages/zone.js/lib/common/error-rewrite.ts#L132

In addition to "cleaning up extra frames", it appends the zone name in square brackets:
https://github.com/angular/angular/blob/3166cffd28e12d882bef3e55472e0a50f4cfc6f5/packages/zone.js/lib/common/error-rewrite.ts#L107-L108

This transforms a frame from:
```
    at Class.method (filpath.ts:line:col)
```
to:
```
    at Class.method (filpath.ts:line:col) [zoneName]
```

Stackdriver Error Reporting has [an endpoint for reporting events](https://cloud.google.com/error-reporting/reference/rest/v1beta1/projects.events/report). In order for an event to be successfully reported, the event must meet at least one of two criteria:
1. The event explicitly contains context.reportLocation, or
2. The event contains a message with a properly formatted stack trace.

For JavaScript, the endpoint expects the message to "be the value of [error.stack](https://github.com/v8/v8/wiki/Stack-Trace-API) as returned by V8."

When sending the zone.js rewritten stack as the message for an event, stackdriver responds with:
```
{
  "error": {
    "code": 400,
    "message": "ReportedErrorEvent.context must contain a location unless `message` contain an exception or stacktrace.",
    "status": "INVALID_ARGUMENT"
  }
}
```

If instead I send error.originalStack as the message, the response is 200 and the event is reported.

While there's no issue here that cannot be worked around, I think it's an unexpected gotcha that you cannot use error.stack in the message, particularly because this process of rewriting the stack isn't particularly easy to find in documentation.

## 🔬 Minimal Reproduction
Available by request. I believe I have already identified the root causes.

## 🔥 Exception or Error
Noted above

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Latest
</code></pre>

**Comment by jelbourn** [comment]
FYI @JiaLiPassion

**Comment by JiaLiPassion** [comment]
@jelbourn , got it.
@4oo, there are two walk arounds.

1. not using `zone-error` package, the `zone-error` package provide several functionalities.
  - rewrite error stack frames to add `zone information`.
  - handle `MyError extends Error` and `new MyError() instanceof MyError` issue.
  - trim most of the `zone related stack traces`.
So if you are not using those functionalities, you can remove `import 'zone.js/dist/zone-error'` from project.

2. if you want to use those features, but don't want to rewrite stack frames, you can set a flag before `import 'zone.js/dist/zone-error'`.
```
(window as any).__Zone_Error_BlacklistedStackFrames_policy = 'disable';
import 'zone.js/dist/zone-error';
```

Thanks!

**Comment by KevinKelchen** [comment]
Hi @JiaLiPassion, 👋

`environment.ts` as generated by the Angular CLI contains this comment related to importing `zone.js/dist/zone-error`:

> This import should be commented out in production mode because it will have a negative impact on performance if an error is thrown.

Is this still the case? Even when disabling the rewriting of stack frames with the proposed `(window as any).__Zone_Error_BlacklistedStackFrames_policy = 'disable';`?

I'm asking because we use Sentry for error tracking and some stack traces aren't very useful due to zone.js.

Thanks!

Kevin

**Comment by JiaLiPassion** [comment]
@KevinKelchen 
Sure, if you are using `(window as any).__Zone_Error_BlacklistedStackFrames_policy = 'disable';`, there will be no performance impact when create new Error object. But `zone-error` will also not bypass any zone.js related stacktraces, in this case, zone-error will only provide one feature to resolve an `Error inheritance issue`.

```
class MyError extends Error {
   constructor() { super();}
}

const myError = new MyError();
console.log('is MyError?', myError instanceof MyError);
// without zone-error it will be false.
```

**Comment by JiaLiPassion** [comment]
Now `zone-error` will not be imported yb default, so we can close this issue for now.
